internal/wclayer/converttobaselayer.go

```diff
@@ -0,0 +1,164 @@
+package wclayer
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+	"strings"
+	"syscall"
+
+	"github.com/Microsoft/hcsshim/internal/hcserror"
+	"github.com/Microsoft/hcsshim/internal/oc"
+	"github.com/Microsoft/hcsshim/internal/safefile"
+	"github.com/Microsoft/hcsshim/internal/winapi"
+	"github.com/pkg/errors"
+	"go.opencensus.io/trace"
+)
+
+var hiveNames = []string{"DEFAULT", "SAM", "SECURITY", "SOFTWARE", "SYSTEM"}
+
+// Ensure the given file exists as an ordinary file, and create a zero-length file if not.
+func ensureFile(path string, root *os.File) error {
+	stat, err := safefile.LstatRelative(path, root)
+	if err != nil && os.IsNotExist(err) {
+		// Do I need to restore directory-modifcation times on the rest of the path?
+		// That doesn't seem necessary since the directory _is_ being modified now.
+		// See `func (w *baseLayerWriter) Close()` for context
+		newFile, err := safefile.OpenRelative(path, root, 0, syscall.FILE_SHARE_WRITE, winapi.FILE_CREATE, 0)
+		if err != nil {
+			return err
+		}
+		return newFile.Close()
+	}
+
+	if err != nil {
+		return err
+	}
+
+	if !stat.Mode().IsRegular() {
+		fullPath := filepath.Join(root.Name(), path)
+		return errors.Errorf("%s has unexpected file mode %s", fullPath, stat.Mode().String())
+	}
+
+	return nil
+}
+
+// safefileMkdirAll works like os.MkdirAll but with safefile
+// TODO: Add this to safefile?
+func safefileMkdirAllRelative(path string, root *os.File) error {
+	pathParts := strings.Split(path, (string)(filepath.Separator))
+	for index := range pathParts {
+
+		partialPath := filepath.Join(pathParts[0 : index+1]...)
+		stat, err := safefile.LstatRelative(partialPath, root)
+
+		if err != nil && os.IsNotExist(err) {
+			if err := safefile.MkdirRelative(partialPath, root); err != nil {
+				return err
+			}
+			continue
+		}
+
+		if err != nil {
+			return err
+		}
+
+		if !stat.IsDir() {
+			fullPath := filepath.Join(root.Name(), partialPath)
+			return errors.Errorf("%s has unexpected file mode %s", fullPath, stat.Mode().String())
+		}
+	}
+
+	return nil
+}
+
+func ensureBaseLayer(root *os.File) (hasUtilityVM bool, err error) {
+	hasUtilityVM = false
+
+	// The base layer registry hives will be copied from here
+	hiveSourcePath := "Files\\Windows\\System32\\config"
+	if err = safefileMkdirAllRelative(hiveSourcePath, root); err != nil {
+		return
+	}
+
+	for _, hiveName := range hiveNames {
+		hivePath := filepath.Join(hiveSourcePath, hiveName)
+		if err = ensureFile(hivePath, root); err != nil {
+			return
+		}
+	}
+
+	stat, err := safefile.LstatRelative(utilityVMFilesPath, root)
+
+	if err != nil && os.IsNotExist(err) {
+		return false, nil
+	}
+
+	if err != nil {
+		return
+	}
+
+	if !stat.Mode().IsDir() {
+		fullPath := filepath.Join(root.Name(), utilityVMFilesPath)
+		return false, errors.Errorf("%s has unexpected file mode %s", fullPath, stat.Mode().String())
+	}
+
+	// Just check that this  exists as a regular file. If it exists but is not a valid registry hive,
```

Comment:
`this  exists`. -_- Fix this next time you touch the code, @TBBle.

Reply:
![Space](https://media1.tenor.com/images/eec5dee1975025897cdd114d49f7d997/tenor.gif)